### ISSUE TITLE
port Terrain to RTSS

### ISSUE THREAD

**Issue by paroj** [OGRECave/ogre#648]
- [x] implement AlphaSplat SRS
- [x] vertex LOD blending
- [x] color map
- [x] offset map
- [x] highp qualifier in RTSS

**Comment by paroj** [comment]
as this is probably the issue with the most likes, this is a draft where I want things to go: #1762

**Comment by tritonas00** [comment]
Any plans to continue this? I think it will fit nicely with RoR + RTSS per pixel lighting and PSSM3

**Comment by paroj** [comment]
I just rebased #1762 and added demo code to get here:
![terrainlight](https://user-images.githubusercontent.com/4941802/193898891-78faeedc-11cd-4e7a-bd53-70139de19a1d.jpg)

basically, you could copy & paste the code of TerrainMaterialGeneratorA in #1762 and rename it to get started. Just as you seem to have done with TerrainPSSMMaterialGenerator.

Currently the RTSS implementation supports multiple lights (albeit they must be hardcoded) but lacks support for:
- layer splatting
- bump & offset mapping
- color map

so if you want to help out, you can start there.

If you can live with only one directional light though, the Terrain shader supports per-pixel lighting and PSSM since the beginning.

What porting difficult is that `SM2Profile` is public API and we need to provide RTSS quirks for all of those options.
Also, we must generate a second, different, technique for the composite map aka Material LOD.

**Comment by tritonas00** [comment]
Great, i will try it ASAP

**Comment by tritonas00** [comment]
Welp, i compiled latest ogre master branch + https://github.com/OGRECave/ogre/pull/1762 applied and migrated to TerrainMaterialGeneratorA (didn't copy the files, just included them directly). Got one massive shadow and a lot of distortion:

![screenshot_2022-10-11_01-48-04_1](https://user-images.githubusercontent.com/2660424/194963687-113881a1-70fe-4281-b3fe-c1b2604b8d24.png)

**Comment by paroj** [comment]
how does `TerrainMaterialGeneratorA` look without #1762?

**Comment by tritonas00** [comment]
ogre master branch + TerrainMaterialGeneratorA + RTSS PSSM3

![screenshot_2022-10-11_15-49-47_1](https://user-images.githubusercontent.com/2660424/195095311-8cf35fa8-3554-418d-a46f-48885d4c1af0.png)

**Comment by paroj** [comment]
> Got one massive shadow and a lot of distortion:

did you still have a directional light there? PSSM needs that to work. And the material generator in #1762 assumes PSSM to be enabled.

Also I just updated #1762 and now only offset mapping is missing feature-wise. Maybe this improves the situation.

> I think this is the best approach

yes, for shipping something using master makes sense. TerrainMaterialGeneratorA got many fixes since you copied it over as  OgreTerrainPSSMMaterialGenerator. One of them must have fixed the shadows for you (they work on my machine regardless).

**Comment by tritonas00** [comment]
Latest Ogre master branch + this PR + TerrainMaterialGeneratorA and updated all media (RTShaderLib, Main)

![screenshot_2022-11-22_03-13-15_1](https://user-images.githubusercontent.com/2660424/203190292-a05ba172-2a6c-418a-88de-2b144a483163.png)

**Comment by paroj** [comment]
which rendersystem is this? Because you got code like:
https://github.com/tritonas00/rigs-of-rods/blob/d5ef84e51830d5b9706e872e309c827550c22144/source/main/terrain/TerrainGeometryManager.cpp#L471-L472
which could cause this, if you did not provide the respective maps.

**Comment by tritonas00** [comment]
OpenGL, Archlinux, intel intergraded gpu 4600

Removing those doesn't make any difference

**Comment by paroj** [comment]
> ![](https://user-images.githubusercontent.com/2660424/203190943-c06b2b0b-423c-4db8-8641-6391c35449ab.png)

this does not look that wrong actually. The lighting is about right. It seems that overflow is happening on some pixels.

Could you attach one of the generated fragment shaders? Those are stored in e.g. `~/.cache/OgreMeshViewer/RTShaderCache/`.

You get the name from Ogre log when it says "RTSS: using ..". You can even edit the shaders on disk to try things out - as long as the C++ parameters dont change, Ogre will use what it finds on disk.

Ideally though, you would provide a renderdoc capture that I could look into. However, renderdoc needs GL3+ - but if you enable RTSS for all materials it should work.

**Comment by tritonas00** [comment]
These are for simple terrain

[simple.zip](https://github.com/OGRECave/ogre/files/10069582/simple.zip)

> as long as the C++ parameters dont change, Ogre will use what it finds on disk.

ah nice

**Comment by tritonas00** [comment]
the one about nhelens mountains
```
#version 460
//-----------------------------------------------------------------------------
//                         PROGRAM DEPENDENCIES
//-----------------------------------------------------------------------------
#define USE_OGRE_FROM_FUTURE
#include <OgreUnifiedShader.h>
#include "FFPLib_Transform.glsl"
#include "SGXLib_NormalMap.glsl"
#include "SGXLib_IntegratedPSSM.glsl"
#include "TerrainSurface.glsl"
#include "SGXLib_PerPixelLighting.glsl"

//-----------------------------------------------------------------------------
//                         GLOBAL PARAMETERS
//-----------------------------------------------------------------------------

SAMPLER2D(globalNormal0, 0);
SAMPLER2D(blendTex1, 1);
SAMPLER2D(difftex2, 2);
SAMPLER2D(normtex3, 3);
SAMPLER2D(difftex4, 4);
SAMPLER2D(normtex5, 5);
SAMPLER2D(difftex6, 6);
SAMPLER2D(normtex7, 7);
SAMPLER2DSHADOW(shadow_map8, 8);
uniform	vec4	uvMul0;
uniform	mat3	normal_matrix;
uniform	vec4	pssm_split_points1;
uniform	vec4	inverse_texture_size8;
uniform	vec4	derived_ambient_light_colour;
uniform	vec4	surface_emissive_colour;
uniform	vec4	derived_scene_colour;
uniform	float	surface_shininess;
uniform	vec4	light_direction_view_space;
uniform	vec4	derived_light_diffuse_colour;
uniform	vec4	derived_light_specular_colour;

//-----------------------------------------------------------------------------
//                         MAIN
//-----------------------------------------------------------------------------
IN(f32vec2	iTexcoord_0, 0)
IN(vec4	iTexcoord_1, 1)
IN(vec3	iTexcoord_2, 2)
void main(void) {
	vec3	lTexcoord_0;
	vec4	lColor_1;
	vec4	diffuseSpec;
	vec3	TSnormal;
	vec4	texTmp;
	vec4	lColor_5;
	vec4	blendWeight1;
	float	lShadowFactor;

	gl_FragColor	=	vec4(1.00000,1.00000,1.00000,1.00000);
	SGX_FetchNormal(globalNormal0, iTexcoord_0, lTexcoord_0);
	FFP_Transform(normal_matrix, lTexcoord_0, lTexcoord_0);
	lColor_5	=	vec4(0.00000,0.00000,0.00000,0.00000);
	blendWeight1	=	texture2D(blendTex1, iTexcoord_0);
	blendTerrainLayer(1.00000, iTexcoord_0, uvMul0.x, normtex3, TSnormal, difftex2, diffuseSpec);
	blendTerrainLayer(blendWeight1.x, iTexcoord_0, uvMul0.y, normtex5, TSnormal, difftex4, diffuseSpec);
	blendTerrainLayer(blendWeight1.y, iTexcoord_0, uvMul0.z, normtex7, TSnormal, difftex6, diffuseSpec);
	transformToTS(TSnormal, normal_matrix, lTexcoord_0);
	lColor_1	=	diffuseSpec.z*vec4(1.00000,1.00000,1.00000,1.00000);
	SGX_ShadowPCF4(shadow_map8, iTexcoord_1, inverse_texture_size8.xy, lShadowFactor);
	gl_FragColor	=	derived_scene_colour;
	SGX_Light_Directional_DiffuseSpecular(lTexcoord_0, iTexcoord_2, light_direction_view_space.xyz, derived_light_diffuse_colour.xyz, derived_light_specular_colour.xyz, surface_shininess, gl_FragColor.xyz, lColor_5.xyz);
	SGX_ApplyShadowFactor_Diffuse(derived_scene_colour, lShadowFactor, gl_FragColor);
	lColor_5	=	lColor_5*lShadowFactor;
	lColor_1	=	gl_FragColor;
	gl_FragColor	=	diffuseSpec*gl_FragColor;
	gl_FragColor.xyz	=	gl_FragColor.xyz+lColor_5.xyz;
}
```

until `gl_FragColor	=	derived_scene_colour;` there is no distortion

![screenshot_2022-11-22_20-01-09_1](https://user-images.githubusercontent.com/2660424/203388210-6eb68454-2364-48db-8c01-74b06bb7bbb4.png)

then when `SGX_Light_Directional_DiffuseSpecular` kicks in:

![screenshot_2022-11-22_20-04-10_1](https://user-images.githubusercontent.com/2660424/203388471-030b17a1-1fef-4d8a-b7e7-085630b79255.png)

**Comment by paroj** [comment]
SGX_Light_Directional_DiffuseSpecular itself should be fine as it is used everywhere.

Can you replace
```glsl
gl_FragColor	=	derived_scene_colour;
```
by 1)
```glsl
gl_FragColor	=	diffuseSpec;
```
2)
```glsl
gl_FragColor	=	vec4(lTexcoord_0/2 + 0.5, 1);
```
comment the lines below and post the results?

**Comment by tritonas00** [comment]
`gl_FragColor	=	diffuseSpec;`

![screenshot_2022-11-22_20-51-15_1](https://user-images.githubusercontent.com/2660424/203397409-9a54a633-a4c7-49d4-baea-86ff6319a1e5.png)

`gl_FragColor	=	vec4(lTexcoord_0/2 + 0.5, 1);`

![screenshot_2022-11-22_20-51-40_1](https://user-images.githubusercontent.com/2660424/203397476-92b2781a-1c40-441e-9470-b750b33159cd.png)

all lines below commented

**Comment by paroj** [comment]
both already show the noise that you see in the final image.

Can you try
```glsl
gl_FragColor	=	texture2D(difftex2, mod (iTexcoord_0*uvMul0.x, 1.0));
```

```glsl
gl_FragColor	=	texture2D(difftex3, mod (iTexcoord_0*uvMul0.y, 1.0));
```

```glsl
gl_FragColor	=	texture2D(difftex4, mod (iTexcoord_0*uvMul0.z, 1.0));
```

**Comment by paroj** [comment]
and 
```vec4	diffuseSpec = vec4_splat(0);```

**Comment by tritonas00** [comment]
```
gl_FragColor	=	texture2D(difftex4, mod (iTexcoord_0*uvMul0.z, 1.0));
```

seems the less noisy, its there though and as soon others kick in like `SGX_ShadowPCF4` it becomes more :(

**Comment by tritonas00** [comment]
Quick video of it: https://streamable.com/3hgwxt

**Comment by tritonas00** [comment]
Another one https://streamable.com/rs9iu3

**Comment by paroj** [comment]
> seems the less noisy, its there though and as soon others kick in like SGX_ShadowPCF4 it becomes more :(

leave the lines below commented out still

**Comment by tritonas00** [comment]
> leave the lines below commented out still

Clean i guess https://streamable.com/revrjs

**Comment by paroj** [comment]
try commenting everything in again and initialise these variables:
https://github.com/OGRECave/ogre/issues/648#issuecomment-1324137437

If thats still broken, we need to check the other layers as described here (screenshot sufficient)
https://github.com/OGRECave/ogre/issues/648#issuecomment-1324135043

**Comment by tritonas00** [comment]
`gl_FragColor	=	texture2D(difftex2, mod (iTexcoord_0*uvMul0.x, 1.0));`

![screenshot_2022-11-22_22-33-36_1](https://user-images.githubusercontent.com/2660424/203415864-bf66c652-16c9-49ea-8971-683e6aeb0b52.png)

`gl_FragColor	=	texture2D(difftex3, mod (iTexcoord_0*uvMul0.y, 1.0));`

![screenshot_2022-11-22_22-34-03_1](https://user-images.githubusercontent.com/2660424/203415922-73c756fa-7fb6-47b2-a40f-e091d8cc5210.png)

`gl_FragColor	=	texture2D(difftex4, mod (iTexcoord_0*uvMul0.z, 1.0));`

![screenshot_2022-11-22_22-34-36_1](https://user-images.githubusercontent.com/2660424/203415974-03ad42f1-d9c8-490f-94f6-822090c677b7.png)

```
vec4	diffuseSpec = vec4_splat(0);
vec3	TSnormal = vec3_splat(0);
```

doesn't make any difference

**Comment by paroj** [comment]
do you have this artifacts in the sample browser demos as well?

**Comment by tritonas00** [comment]
> do you have this artifacts in the sample browser demos as well?

I can't compile the samples currently

```
In file included from /home/babis/Downloads/ror-dependencies/Source/ogre/Samples/Simple/src/VolumeCSG.cpp:28:
/home/babis/Downloads/ror-dependencies/Source/ogre/Samples/Simple/include/VolumeCSG.h:41:10: fatal error: OgreVolumeChunk.h: No such file or directory
   41 | #include "OgreVolumeChunk.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[5]: *** [Samples/CMakeFiles/DefaultSamples.dir/build.make:524: Samples/CMakeFiles/DefaultSamples.dir/Simple/src/VolumeCSG.cpp.o] Error 1
make[5]: *** Waiting for unfinished jobs....
In file included from /home/babis/Downloads/ror-dependencies/Source/ogre/Samples/Simple/src/VolumeTerrain.cpp:28:
/home/babis/Downloads/ror-dependencies/Source/ogre/Samples/Simple/include/VolumeTerrain.h:41:10: fatal error: OgreVolumeChunk.h: No such file or directory
   41 | #include "OgreVolumeChunk.h"
```

I will try when i manage to compile them

**Comment by only-a-ptr** [comment]
I rebased #1762 on master, generate VS2022 project, built Debug and tried Terrain sample.
With dx11 I got a segfault - screenshot and stacktrace attached.
GL works but shadows glitch - screenshots attached.
[ogre-648.zip](https://github.com/OGRECave/ogre/files/10071514/ogre-648.zip)

**Comment by paroj** [comment]
> I will try when i manage to compile them

it seems you did not enable the Volume component

To verify my suspicion, you could also use the stock shader again, but replace `uvMul0.x`,`uvMul0.y``uvMul0.z` by `1.0`

**Comment by paroj** [comment]
> GL works but shadows glitch - screenshots attached. Also I noticed the houses don't receive color shadows 

the shadow glitch is caused by the PSSM being confused with the second light. This is unrelated to the change here - as in it would also happen with Ogre 1.11.
The houses not receiving color shadows is how color shadows work.
TLDR; all your GL screenshots are correct as far as this issue goes.

D3D11 needs some more testing as I develop on GL first.

**Comment by tritonas00** [comment]
> To verify my suspicion, you could also use the stock shader again, but replace uvMul0.x,uvMul0.y,uvMul0.z by 1.0

This produces:

![screenshot_2022-11-23_15-48-13_1](https://user-images.githubusercontent.com/2660424/203566343-8dac8fd2-4150-4db5-94ea-0fc8c5bd5fc9.png)

> it seems you did not enable the Volume component

yep, noise there also

![kk](https://user-images.githubusercontent.com/2660424/203566457-1789e06c-063e-4291-bd15-a82084905bd0.png)

**Comment by paroj** [comment]
It seems to be something specific to your GPU, as neither me nor @only-a-ptr  have this issue in the Sample.
I even tried it with GLES2 on my phone and it worked correctly.

This is getting hard to debug now. At least the old shaders did work for you.

Do you have this issue in the "Endless World" Sample too?

And can you go back to this commit and check? https://github.com/OGRECave/ogre/pull/1762/commits/e5ef04fd4360b2960746b75c7c28692ae5d62522
It only includes the minor algorithmic shader changes, while still using the old shader generator.

**Comment by tritonas00** [comment]
> Do you have this issue in the "Endless World" Sample too?

yep, with both renderers

![kk](https://user-images.githubusercontent.com/2660424/203575224-f73e161c-f17e-432d-8a17-3e664359ce2b.png)

> And can you go back to this commit and check? e5ef04f

will try that later

**Comment by tritonas00** [comment]
> And can you go back to this commit and check? https://github.com/OGRECave/ogre/commit/e5ef04fd4360b2960746b75c7c28692ae5d62522

This is noise free 

![screenshot_2022-11-23_17-47-07_1](https://user-images.githubusercontent.com/2660424/203589552-806f2776-eb65-4c95-a0c6-94d501f3ebe8.png)

**Comment by paroj** [comment]
the issue is somewhere here then:
old:
```glsl
void main()
{
    float shadow = 1.0;
    vec2 uv = oUVMisc.xy;
    gl_FragColor = vec4(0,0,0,1);
    vec3 normal;
    SGX_FetchNormal(globalNormal, uv, normal);
    normal = ((normalMatrix) * (normal));
    vec3 lightDir = -lightPos.xyz;
    vec3 viewPos = oPosObj.xyz;
    vec4 diffuseSpec = vec4(0.0);
    vec4 blendTexVal0 = texture2D(blendTex0, uv);
    vec3 TSnormal = vec3(0.0);
    blendTerrainLayer(1.0, uv, uvMul_0.r, viewPos, scaleBiasSpecular.xy, normtex0, TSnormal, difftex0, diffuseSpec);
    transformToTS(TSnormal, normalMatrix, normal);

    vec3 specularCol = vec3(0,0,0);
    SGX_Light_Directional_DiffuseSpecular(normal, viewPos, lightDir, lightDiffuseColour * diffuseSpec.rgb, lightSpecularColour * diffuseSpec.a, scaleBiasSpecular.z, gl_FragColor.rgb, specularCol);
    gl_FragColor.rgb += specularCol;
    gl_FragColor.rgb = gl_FragColor.rgb * shadow + ambient.rgb * diffuseSpec.rgb;
    gl_FragColor.rgb = mix(fogColour, gl_FragColor.rgb, fogVal);
}
```

new:
```glsl
void main(void) {
	vec3	lTexcoord_0;
	vec4	lColor_1;
	vec4	diffuseSpec;
	vec3	TSnormal;
	vec4	texTmp;
	vec4	lColor_5;
	vec4	blendWeight1;

	FragColor	=	vec4(1.00000,1.00000,1.00000,1.00000);
	SGX_FetchNormal(globalNormal0, iTexcoord_0, lTexcoord_0);
	FFP_Transform(normal_matrix, lTexcoord_0, lTexcoord_0);
	lColor_5	=	vec4(0.00000,0.00000,0.00000,0.00000);
	blendWeight1	=	texture(blendTex1, iTexcoord_0);
	diffuseSpec	=	vec4(0.00000,0.00000,0.00000,0.00000);
	TSnormal	=	vec3(0.00000,0.00000,0.00000);
	blendTerrainLayer(1.00000, iTexcoord_0, uvMul0.x, iTexcoord_1, vec2(0.0300000,-0.0400000), normtex3, TSnormal, difftex2, diffuseSpec);
	transformToTS(TSnormal, normal_matrix, lTexcoord_0);
	lColor_1	=	diffuseSpec.z*vec4(1.00000,1.00000,1.00000,1.00000);
	FragColor	=	derived_scene_colour;
	SGX_Light_Directional_DiffuseSpecular(lTexcoord_0, iTexcoord_1, light_direction_view_space.xyz, derived_light_diffuse_colour.xyz, derived_light_specular_colour.xyz, surface_shininess, FragColor.xyz, lColor_5.xyz);
	lColor_1	=	FragColor;
	FragColor	=	diffuseSpec*FragColor;
	FragColor.xyz	=	FragColor.xyz+lColor_5.xyz;
	FragColor	=	mix(fog_colour, FragColor, iTexcoord_2);
}
```
these shaders are from EndlessWorld as it is the most simple case: 1 layer, 1 light, no other objects.

As I cannot do much more than staring at this, you will have to investigate. As EndlessWorld reproduces this, I suggest using https://renderdoc.org/ on SampleBrowser using GL3+.

You can store a capture where it works and one where it does not. Although, I doubt it will be broken for me, you could upload the broken capture here.

In the pipeline view, you can live-edit the shader to write whatever you want to FragColor. We already narrowed it down to being broken after the `blendTerrainLayer` call.
You will now have to open both captures and compare the intermediate results until you find what is causing the noise there.

**Comment by tritonas00** [comment]
Ok i build and enabled GL3PLUS (previously i tested with OpenGL and OpenGL 2 options) . No issues at all there, both Endless World and Terrain, 1762 fully merged

**Comment by AnotherFoxGuy** [comment]
> So i guess we need to move RoR to DX11/GL3+ ?

I would prefer to move to Vulcan, but I don't know what the implementation status of it is in OGRE 13+

**Comment by paroj** [comment]
> OpenGL and OpenGL 2 options

what is OpenGL 2? Do you mean OpenGL ES 2.x?

> So i guess we need to move RoR to DX11/GL3+ ?

I mean this bug is still solvable. Whatever is easier for you I guess.

**Comment by tritonas00** [comment]
> what is OpenGL 2? Do you mean OpenGL ES 2.x?

OpenGL ES 2.x, sorry.

> I mean this bug is still solvable. Whatever is easier for you I guess.

But how to debug? i can't reproduce with GL3+/renderdoc so i can provide captures

**Comment by paroj** [comment]
> But how to debug?

you could still edit the on-disk shaders for GL and compare against a GL3+ RenderDoc capture. It would be only slightly more painful.

Out of curiosity, what happens if you disable the Fixed Function Pipeline on GL?

> OpenGL ES 2.x, sorry.

and GLES had the same issue? If so, you can use renderdoc on that.

**Comment by tritonas00** [comment]
Hm ok i need to delete my ~.cache/OGRE Sample Browser folder and start fresh with each render system, because the setting doesn't seem to change at all after. So:

OpenGL doesn't have the issue
OpenGL ES 2.x  does have the issue

**Comment by paroj** [comment]
> Update: with vanilla master, dx11 works but the terrain is extremely grainy. 

looked into this. The cause is that D3D11 cannot auto-generate mipmaps for compressed textures. The grain is simply aliasing.
The same applies to GLES2.

**Comment by paroj** [comment]
> OpenGL doesn't have the issue
OpenGL ES 2.x does have the issue

@tritonas00 I thought the whole point was that OpenGL had the issues in RoR? Or did you mean GL3+ there?

**Comment by tritonas00** [comment]
> @tritonas00 I thought the whole point was that OpenGL had the issues in RoR? Or did you mean GL3+ there?

Was talking about the samples. RoR uses OpenGL only and has the issue. Samples with OpenGL doesnt have the issue, strangely. Samples with OpenGL ES 2.x does have the issue. At least on my machine.

**Comment by paroj** [comment]
@tritonas00 as said in the previous comment, the difference between GL and GLES is this:

```
// GL
Texture 'Ground37_diffspec.dds': Loading 1 faces(PF_DXT5,1024x1024x1) with 10 hardware generated mipmaps from Image. Internal format is PF_DXT5,1024x1024x1.
// GLES2
Texture 'Ground37_diffspec.dds': Loading 1 faces(PF_DXT5,1024x1024x1) with 0 custom mipmaps from Image. Internal format is PF_DXT5,1024x1024x1.
```

so at least for the SampleBrowser the mystery is solved. You could verify whether there are no mipmaps in your RoR levels either.

**Comment by tritonas00** [comment]
There are

```
16:04:43: Texture 'a1da0uid-nhelens.dds': Loading 1 faces(PF_DXT5,1024x1024x1) with 10 custom mipmaps from Image. Internal format is PF_DXT5,1024x1024x1.
16:04:43: Texture 'blank_NRM.dds': Loading 1 faces(PF_A8R8G8B8,800x600x1) with 5 hardware generated mipmaps from Image. Internal format is PF_A8R8G8B8,800x600x1.
16:04:43: Texture 'nhelens_terrain_detail2.dds': Loading 1 faces(PF_DXT5,512x512x1) with 9 custom mipmaps from Image. Internal format is PF_DXT5,512x512x1.
16:04:43: Texture 'nhelens_terrain_detail1.dds': Loading 1 faces(PF_DXT5,512x512x1) with 9 custom mipmaps from Image. Internal format is PF_DXT5,512x512x1.
```

**Comment by tritonas00** [comment]
RTSS Terrain sample renderdoc frame captures

GL ES 2.x, it has the noise issue
https://www.dropbox.com/s/20awgodkcwt4woc/issue.rdc?dl=1

GL3+, it doesn't have the noise issue
https://www.dropbox.com/s/1oh38oes22igyvv/no-issue.rdc?dl=1

**Comment by paroj** [comment]
> GL ES 2.x, it has the noise issue

the capture renders like this on my machine
![issue](https://user-images.githubusercontent.com/4941802/203855392-fe48b816-da56-4212-af56-9aa38fdedbd1.jpg)

this is definately a mipmapping issue.

> There are (log from Ogre 1.11.6)

yes, there are mipmaps and it works. it would be more interesting to see the log for when you have the noise.

to make it double sure, you could convert the DDS texture to PNG (and change paths in code), then it should work.
```sh
# e.g. with imagick
convert some.dds some.png
```

**Comment by tritonas00** [comment]
> the capture renders like this on my machine

yep, thats it. Imagine that x5 or so in RoR, depending the terrain (depending the mipmap levels?)

> it would be more interesting to see the log for when you have the noise.

here is the full log, you can start from `LOADING TERRAIN a1da0UID-nhelens.terrn2`
[RoR.log](https://github.com/OGRECave/ogre/files/10087705/RoR.log)
```
01:56:59: RTSS: using 7ee6a90b3496e6492ea32b1fe756b339_VS for Pass 0 of 'OgreTerrain/3440999213'
01:56:59: RTSS: using 8e36b92e4ac807edb9f0a67630a8dcdc_FS for Pass 0 of 'OgreTerrain/3440999213'
01:56:59: RTSS: using d2c02dc16b5a833740e663aa969b68d6_VS for Pass 0 of 'OgreTerrain/3440999213'
01:56:59: RTSS: using d04d8a480e9a739b8e0bab712fbc8afe_FS for Pass 0 of 'OgreTerrain/3440999213'
01:56:59: Texture 'a1da0uid-nhelens.dds': Loading 1 faces(PF_DXT5,1024x1024x1) with 5 custom mipmaps from Image. Internal format is PF_DXT5,1024x1024x1.
01:56:59: Texture 'blank_NRM.dds': Loading 1 faces(PF_A8R8G8B8,800x600x1) with 5 hardware generated mipmaps from Image. Internal format is PF_A8R8G8B8,800x600x1.
01:56:59: Texture 'nhelens_terrain_detail2.dds': Loading 1 faces(PF_DXT5,512x512x1) with 5 custom mipmaps from Image. Internal format is PF_DXT5,512x512x1.
01:56:59: Texture 'nhelens_terrain_detail1.dds': Loading 1 faces(PF_DXT5,512x512x1) with 5 custom mipmaps from Image. Internal format is PF_DXT5,512x512x1.
01:56:59: RTSS: using 84c116655605cf8b342625f2b413044f_VS for Pass 0 of 'OgreTerrain/3440999213/comp'
01:56:59: RTSS: using dffe3523372a7a710c13d2a3fb2c2834_FS for Pass 0 of 'OgreTerrain/3440999213/comp'
01:56:59: RTSS: using 7ee6a90b3496e6492ea32b1fe756b339_VS for Pass 0 of 'OgreTerrain/3440999213'
01:56:59: RTSS: using 8e36b92e4ac807edb9f0a67630a8dcdc_FS for Pass 0 of 'OgreTerrain/3440999213'
01:56:59: RTSS: using d2c02dc16b5a833740e663aa969b68d6_VS for Pass 0 of 'OgreTerrain/3440999213'
01:56:59: RTSS: using d04d8a480e9a739b8e0bab712fbc8afe_FS for Pass 0 of 'OgreTerrain/3440999213'
01:56:59: RTSS: using 84c116655605cf8b342625f2b413044f_VS for Pass 0 of 'OgreTerrain/3440999213/comp'
01:56:59: RTSS: using dffe3523372a7a710c13d2a3fb2c2834_FS for Pass 0 of 'OgreTerrain/3440999213/comp'
01:57:00: done loading page: loaded 3 layers
```

> to make it double sure, you could convert the DDS texture to PNG (and change paths in code), then it should work.

will give it a try

**Comment by tritonas00** [comment]
Tried with png, same noise issue

```
02:21:26: Texture 'a1da0uid-nhelens.png': Loading 1 faces(PF_A8R8G8B8,1024x1024x1) with 5 hardware generated mipmaps from Image. Internal format is PF_A8R8G8B8,1024x1024x1.
02:21:26: Texture 'blank_NRM.png': Loading 1 faces(PF_A8R8G8B8,800x600x1) with 5 hardware generated mipmaps from Image. Internal format is PF_A8R8G8B8,800x600x1.
02:21:26: Texture 'nhelens_terrain_detail2.png': Loading 1 faces(PF_A8R8G8B8,512x512x1) with 5 hardware generated mipmaps from Image. Internal format is PF_A8R8G8B8,512x512x1.
02:21:26: Texture 'nhelens_terrain_detail1.png': Loading 1 faces(PF_A8R8G8B8,512x512x1) with 5 hardware generated mipmaps from Image. Internal format is PF_A8R8G8B8,512x512x1.
```

**Comment by tritonas00** [comment]
So to summarize my findings on my machine:

RTSS Terrain sample
GL -> no noise
GLES -> noise
GL3+ -> no noise

RoR
GL -> noise

**Comment by tritonas00** [comment]
Ok managed to run RoR with GL3+. Noise free

![screenshot_2022-11-25_03-56-57_1](https://user-images.githubusercontent.com/2660424/203885025-baeb2fbd-5d82-4c5c-80c2-93e66adcc837.png)

![screenshot_2022-11-25_03-58-04_1](https://user-images.githubusercontent.com/2660424/203885088-c02be6e5-d4c5-4f15-92f2-a5b416c4622a.png)

**Comment by CuriousMike56** [comment]
@tritonas00 told me to post here. Terrain sample result on all render  systems (1660ti):
![image](https://user-images.githubusercontent.com/46073351/203898269-88a40ad1-3ee5-4bd5-960f-f12a38bad4a6.png)
In RoR I get the same result as Tritonas using both DirectX9 and OpenGL:
![image](https://user-images.githubusercontent.com/46073351/203898224-e263ecc9-ed53-4d3e-9fd4-30da001bee63.png)

**Comment by tritonas00** [comment]
It seems my gpu doesn't play well with RTSS terrain + OpenGL i guess  ¯\_(ツ)_/¯

**Comment by paroj** [comment]
> Ok managed to run RoR with GL3+. Noise free

but it is texture free also. We already found out that the noise comes from texture filtering.

Anything in the logs why the textures are gone?

**Comment by tritonas00** [comment]
> RoR
GL -> noise
GL3+ -> no noise

I compared the shaders and it seems the difference is (FS)
GL
```
uniform	vec4	uvMul0;
uniform	mat3	normal_matrix;
uniform	vec4	pssm_split_points1;
uniform	vec4	inverse_texture_size8;
uniform	vec4	derived_ambient_light_colour;
uniform	vec4	surface_emissive_colour;
uniform	vec4	derived_scene_colour;
uniform	float	surface_shininess;
uniform	vec4	light_direction_view_space;
uniform	vec4	derived_light_diffuse_colour;
uniform	vec4	derived_light_specular_colour;
```

GL3+
```
layout(location = 0) uniform	vec4	uvMul0;
layout(location = 1) uniform	mat3	normal_matrix;
layout(location = 4) uniform	vec4	pssm_split_points1;
layout(location = 5) uniform	vec4	inverse_texture_size8;
layout(location = 6) uniform	vec4	derived_ambient_light_colour;
layout(location = 7) uniform	vec4	surface_emissive_colour;
layout(location = 8) uniform	vec4	derived_scene_colour;
layout(location = 9) uniform	float	surface_shininess;
layout(location = 10) uniform	vec4	light_direction_view_space;
layout(location = 11) uniform	vec4	derived_light_diffuse_colour;
layout(location = 12) uniform	vec4	derived_light_specular_colour;
```

Also in VS
GL
```
uniform	mat4	worldviewproj_matrix;
uniform	vec2	custom1001;
uniform	mat4	texture_worldviewproj_matrix;
uniform	mat3	normal_matrix;
uniform	mat4	worldview_matrix;
```

GL3+
```
layout(location = 0) uniform	mat4	worldviewproj_matrix;
layout(location = 4) uniform	vec2	custom1001;
layout(location = 5) uniform	mat4	texture_worldviewproj_matrix;
layout(location = 9) uniform	mat3	normal_matrix;
layout(location = 12) uniform	mat4	worldview_matrix;
```

> Anything in the logs why the textures are gone?

Can't see anything
[RoR.log](https://github.com/OGRECave/ogre/files/10093494/RoR.log)
